### PR TITLE
Consistent use of DTOs in Jira/Bitbucket/BambooService and improved test coverage in BitbucketService

### PR DESCRIPTION
**Pull request by derLalla**
### Description

This add test cases for the BitbucketService and handles edge cases by catching Exceptions.
The methods deleteWebHook() and deleteExistingWebHooks() were unused and therefore deleted.

Also this PR gets rid of the Map Casting in the restTemplate and its mock functions by replacing them with DTOs.

### Steps for Testing

As now all rest calls for Bamboo/Bitbucket/Jira are called with DTOs, all relevant rest calls should be tested. 

1. Login using a TUM account over JIRA
2. Create a Programming Exercise and verify that all necessary steps are working as expected.
3. Participate in it and also verify that all works as expected.
4. Delete the exercise and make sure everything works
5. Use an admin account to change the groups of an existing user in the user management view

### Test Coverage

* BitbucketService.java: 51% -> 76%

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/BitbucketService.java [hunk 1]
```diff
@@ -599,9 +583,10 @@ private void grantGroupPermissionToProject(String projectKey, String groupName,
      */
     private Map<Integer, String> getExistingWebHooks(String projectKey, String repositorySlug) throws BitbucketException {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/BitbucketService.java:584]: we should return `List<BitbucketWebHookDTO>` here and simply return `response.getBody().getSearchResults()` or `List.of()`

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketCloneDTO.java [hunk 2]
```diff
@@ -30,6 +30,18 @@ public void setCloneDetails(CloneDetailsDTO cloneDetails) {
         this.cloneDetails = cloneDetails;
     }
 
+    /**
+     * Creating a description for the console log
+     * @return description with name and project key
+     */
+    public String toString() {
+        String description = String.format("name=%s", name);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketCloneDTO.java:38]: please avoid String.format and use normal string concatenation

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketCloneDTO.java [hunk 3]
```diff
@@ -45,5 +57,9 @@ public String getKey() {
         public void setKey(String key) {
             this.key = key;
         }
+
+        public String toString() {
+            return String.format("project={key=%s}", key);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketCloneDTO.java:62]: please avoid String.format and use normal string concatenation
check out other toString() methods, you can actually generate those automatically in IntelliJ

**Review by krusche — CHANGES_REQUESTED** [review]
I just skimmer over the changes for the DTOs. They look good. Some improvements are still possible and it would be great if you could also convert the latest places in BambooService into DTOs

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/BitbucketService.java (hunk 1 above)
Comment (derLalla) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/BitbucketService.java:584]: good idea. I've fixed that

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketCloneDTO.java (hunk 3 above)
Comment (derLalla) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketCloneDTO.java:62]: thanks for the tip with IntelliJ

**Review by tiffca — CHANGES_REQUESTED** [review]
I'm not able to access any of the admin menu items. I compared this to the develop branch where I am able to, so I think it's something related to this pr. 

Steps to reproduce (test server 1) 
1. As a TUM account user and an admin user, log in 
2. Try to access any of the admin menu items, such as user management.  
3. The menus are not accessible. Expected: should be able to access the admin menus 



logs: 
bootstrap:128 GET https://artemistest.ase.in.tum.de/app/13.20dfd72fed667222a82d.chunk.js net::ERR_ABORTED 404
r.e @ bootstrap:128
loadChildren @ app-routing.module.ts:15
loadModuleFactory @ router.js:3493
load @ router.js:3479
Show 52 more frames
instrument.js:109 ERROR Error: Uncaught (in promise): ChunkLoadError: Loading chunk 13 failed.
(error: https://artemistest.ase.in.tum.de/app/13.20dfd72fed667222a82d.chunk.js)
ChunkLoadError: Loading chunk 13 failed.
(error: https://artemistest.ase.in.tum.de/app/13.20dfd72fed667222a82d.chunk.js)
    at Function.r.e (bootstrap:105)
    at loadChildren (app-routing.module.ts:15)
    at te.loadModuleFactory (router.js:3493)
    at te.load (router.js:3479)
    at l.project (router.js:2441)
    at l._tryNext (mergeMap.js:44)
(anonymous) @ instrument.js:109
Wo @ core.js:4352
handleError @ core.js:4400
handleError @ sentry.error-handler.ts:51
next @ core.js:28039
o @ core.js:24853

**Comment by derLalla** [comment]
> I'm not able to access any of the admin menu items. I compared this to the develop branch where I am able to, so I think it's something related to this pr. Programming tests worked fine!
> 
> Steps to reproduce (test server 1)
> 
> 1. As a TUM account user and an admin user, log in (I tested with my own user/TUM identifier)
> 2. Try to access any of the admin menu items, such as user management.
> 3. The menus are not accessible. Expected: should be able to access the admin menus
> 
> 
> 
> logs:
> bootstrap:128 GET https://artemistest.ase.in.tum.de/app/13.20dfd72fed667222a82d.chunk.js net::ERR_ABORTED 404
> r.e @ bootstrap:128
> loadChildren @ app-routing.module.ts:15
> loadModuleFactory @ router.js:3493
> load @ router.js:3479
> Show 52 more frames
> instrument.js:109 ERROR Error: Uncaught (in promise): ChunkLoadError: Loading chunk 13 failed.
> (error: https://artemistest.ase.in.tum.de/app/13.20dfd72fed667222a82d.chunk.js)
> ChunkLoadError: Loading chunk 13 failed.
> (error: https://artemistest.ase.in.tum.de/app/13.20dfd72fed667222a82d.chunk.js)
> at Function.r.e (bootstrap:105)
> at loadChildren (app-routing.module.ts:15)
> at te.loadModuleFactory (router.js:3493)
> at te.load (router.js:3479)
> at l.project (router.js:2441)
> at l._tryNext (mergeMap.js:44)
> (anonymous) @ instrument.js:109
> Wo @ core.js:4352
> handleError @ core.js:4400
> handleError @ sentry.error-handler.ts:51
> next @ core.js:28039
> o @ core.js:24853

I've tested it on testserver1 with my tum credentials and I don't have this problem. I have access to all menu entries

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]
https://sentry.io/organizations/ls1intum/issues/1986426568/?project=1440029&query=is%3Aunresolved

Had this error when starting the programming exercise. Did not happen on the second try to start the programming exercise.

image

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketCloneDTO.java [hunk 4]
```diff
@@ -30,6 +36,15 @@ public void setCloneDetails(CloneDetailsDTO cloneDetails) {
...
+    /**
+     * Creating a description for the console log
+     * @return description with name and project key
+     */
+    @Override
+    public String toString() {
+        return "BitbucketCloneDTO{" + "name='" + name + '\'' + ", cloneDetails=" + cloneDetails.toString() + '}';
+    }
+
     public static final class CloneDetailsDTO {
```
Comment (fde312) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketCloneDTO.java:49]: Maybe add here also a constructor, like for `BitbucketCloneDTO` :)

**Review by fde312 — CHANGES_REQUESTED** [review]
We found a issue: 
image

Could be related to the inline comment I left you :)

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketCloneDTO.java (hunk 4 above)
Comment (derLalla) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketCloneDTO.java:49]: yes thanks!

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]
Error from before was fixed (Ignore this change request)

**Review by stefanwaldhauser — DISMISSED** [review]
Everything worked in the test session now, and code looks finde. I approve 👍🏻

**Review by fde312 — DISMISSED** [review]
Works fine and code looks also good! Good job! 👍

**Review by kloessst — CHANGES_REQUESTED** [review]
Tested:
- Programming exercise creation
- Import of programming exercise
- Participation
- Exam mode with programming exercises
- Removing a group from a user via user management

All worked fine! One small note though:

**Review by tiffca — DISMISSED** [review]
Re-tested on test server 1 and I can't reproduce the admin error, I was able to remove/add groups from a user. Nice job!

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java [hunk 5]
```diff
@@ -496,11 +496,10 @@ public void updatePlanRepository(String bambooProject, String buildPlanKey, Stri
...
     @Override
     public String getPlanKey(Object requestBody) throws BambooException {
-        // TODO: convert into a proper DTO object to avoid unchecked Map casts
         try {
-            Map<String, Object> requestBodyMap = (Map<String, Object>) requestBody;
-            Map<String, Object> planMap = (Map<String, Object>) requestBodyMap.get("plan");
-            return (String) planMap.get("key");
+            final var buildResult = mapper.convertValue(requestBody, BambooBuildResultNotificationDTO.class);
+            var plan = buildResult.getPlan();
+            return plan.getKey();
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java:502]:
```suggestion
            return buildResult.getPlan().getKey();
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java [hunk 6]
```diff
@@ -496,11 +496,10 @@ public void updatePlanRepository(String bambooProject, String buildPlanKey, Stri
      */
     @Override
     public String getPlanKey(Object requestBody) throws BambooException {
-        // TODO: convert into a proper DTO object to avoid unchecked Map casts
         try {
-            Map<String, Object> requestBodyMap = (Map<String, Object>) requestBody;
-            Map<String, Object> planMap = (Map<String, Object>) requestBodyMap.get("plan");
-            return (String) planMap.get("key");
+            final var buildResult = mapper.convertValue(requestBody, BambooBuildResultNotificationDTO.class);
+            var plan = buildResult.getPlan();
```
[the remark this answers is not in the document]
Comment (derLalla) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java]: well seen. I've changed it.

**Review by kloessst — APPROVED** [review]
Works fine!

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketWebHookDTO.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    public String getUrl() {
+        return url;
+    }
+
+    public void setUrl(String url) {
+        this.url = url;
+    }
+
+    public ArrayList<String> getEvents() {
```
[the remark this answers is not in the document]
Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketWebHookDTO.java:42]: @derLalla I actually found that warning useful - Can't we use List here?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketBranchProtectionDTO.java [hunk 8]
```diff
@@ -9,6 +9,12 @@
 
     private MatcherDTO matcher;
 
+    /**
+     * needed for Jackson
+     */
+    public BitbucketBranchProtectionDTO() {
```
[the remark this answers is not in the document]
Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketBranchProtectionDTO.java:15]: @derLalla I believe you could prevent these warning if you put the documentation inside the (empty) constructor instead of above, in case you are interested in changing this.

Otherwise just resolve it again.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketWebHookDTO.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (MaisiKoleni) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/bitbucket/dto/BitbucketWebHookDTO.java:42]: Please use `List` here.

File: src/test/java/de/tum/in/www1/artemis/AbstractSpringIntegrationJenkinsGitlabTest.java [hunk 9]
```diff
@@ -89,7 +90,13 @@ public void mockConnectorRequestsForSetup(ProgrammingExercise exercise) throws E
     }
 
     @Override
-    public List<Verifiable> mockConnectorRequestsForStartParticipation(ProgrammingExercise exercise, String username, Set<User> users) throws IOException, URISyntaxException {
+    public void mockCopyRepositoryForParticipation(ProgrammingExercise exercise, String username, HttpStatus status) throws URISyntaxException, IOException {
```
Comment (anditurdiu) [inline src/test/java/de/tum/in/www1/artemis/AbstractSpringIntegrationJenkinsGitlabTest.java:93]: The codacy point here I believe is valid. If you make this method abstract you can ensure that it is not called accidentally while it is not implemented.

File: src/test/java/de/tum/in/www1/artemis/AbstractSpringIntegrationJenkinsGitlabTest.java [hunk 10]
```diff
@@ -89,7 +90,13 @@ public void mockConnectorRequestsForSetup(ProgrammingExercise exercise) throws E
...
+        // TODO: implement
+    }
+
+    @Override
+    public List<Verifiable> mockConnectorRequestsForStartParticipation(ProgrammingExercise exercise, String username, Set<User> users, boolean ltiUserExists)
+            throws IOException, URISyntaxException {
```
Comment (anditurdiu) [inline src/test/java/de/tum/in/www1/artemis/AbstractSpringIntegrationJenkinsGitlabTest.java:99]: Same here: An empty method in an abstract class should be abstract instead

File: src/test/java/de/tum/in/www1/artemis/connector/bitbucket/BitbucketRequestMockProvider.java [hunk 11]
```diff
@@ -146,24 +159,59 @@ public void mockGetBitbucketRepository(ProgrammingExercise exercise, String bitb
...
-    public void mockConfigureRepository(ProgrammingExercise exercise, String username, Set<User> users) throws URISyntaxException, IOException {
+    public void mockConfigureRepository(ProgrammingExercise exercise, String username, Set<User> users, boolean ltiUserExists) throws URISyntaxException, IOException {
         final var projectKey = exercise.getProjectKey();
         final var repoName = projectKey.toLowerCase() + "-" + username.toLowerCase();
         for (User user : users) {
             if (exercise.hasCourse()) {
-                mockGiveWritePermission(exercise, repoName, user.getLogin());
+                // add mock for userExists() check, if the username contains edx_ or u4i_
+                String loginName = user.getLogin();
+                if (userPrefixEdx.isPresent() && loginName.startsWith(userPrefixEdx.get()) || userPrefixU4I.isPresent() && loginName.startsWith((userPrefixU4I.get()))) {
```
Comment (anditurdiu) [inline src/test/java/de/tum/in/www1/artemis/connector/bitbucket/BitbucketRequestMockProvider.java:169]:
```suggestion
                if (userPrefixEdx.isPresent() && loginName.startsWith(userPrefixEdx.get()) || userPrefixU4I.isPresent() && loginName.startsWith(userPrefixU4I.get())) {
```

**Review by anditurdiu — CHANGES_REQUESTED** [review]
Code looks good generally just some minor suggestions. There are also comments from @sleiss & @MaisiKoleni which are not implemented yet

File: src/test/java/de/tum/in/www1/artemis/AbstractSpringIntegrationJenkinsGitlabTest.java (hunk 9 above)
Comment (krusche) [reply src/test/java/de/tum/in/www1/artemis/AbstractSpringIntegrationJenkinsGitlabTest.java:93]: This is intentional. This is a requirement by the interface that is implemented. This method has to be implemented later on, when we implement the tests for Jenkins

File: src/test/java/de/tum/in/www1/artemis/AbstractSpringIntegrationJenkinsGitlabTest.java (hunk 10 above)
Comment (krusche) [reply src/test/java/de/tum/in/www1/artemis/AbstractSpringIntegrationJenkinsGitlabTest.java:99]: This is intentional. This is a requirement by the interface that is implemented. This method has to be implemented later on, when we implement the tests for Jenkins

**Review by MaisiKoleni — APPROVED** [review]
Works still well, tested on test server 3.

**Review by sleiss — APPROVED** [review]
Code looks fine so far, have not tested it.

**Review by kloessst — APPROVED** [review]
Re-approve

**Comment by anditurdiu** [comment]
@fde312 and I found an issue with regards to test runs, sometimes you cannot switch between programming exercises. We are not sure if this is related to the changes introduced here or to the big TypeScript PR we did a while ago, however we could not reproduce this on other branches.
